src/OpenTelemetry.Extensions.Hosting/Implementation/EventSourceEventFormatting.cs

```diff
@@ -0,0 +1,93 @@
+// <copyright file="EventSourceEventFormatting.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics.Tracing;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+
+namespace OpenTelemetry.Internal
+{
+    internal static class EventSourceEventFormatting
+    {
+        private static readonly StringBuilder Builder = new StringBuilder();
+
+        public static string Format(EventWrittenEventArgs eventData)
+        {
+            var payloadArray = eventData.Payload?.ToArray() ?? Array.Empty<object>();
```

Comment:
I feel like this array allocation + copy could be avoided. Payload is a ReadOnlyCollection<object> already I think? Could we do...

```csharp
private static readonly ReadOnlyCollection<object> emptyPayload = new ReadOnlyCollection<object>(Array.Empty<object>());

var payloadCollection = eventData.Payload ?? emptyPayload;
```

I think the for loops don't need to change, there should be count + indexer on ReadOnlyCollection.

Reply:
The items are being formatted and updated in `ProcessPayloadArray` so it can't be a read only collection.